### PR TITLE
Framework: Add CSS modules support

### PR DESCRIPTION
**Pull request by stephanethomas**
This pull request is a follow-up of #10366 that seeks to introduce [CSS modules](https://github.com/css-modules/css-modules) to Calypso. CSS modules ([here](https://css-tricks.com/css-modules-part-1-need/) is a good introduction) bring numerous advantages, the main one being that we no longer need to name classes [using a BEM-like syntax](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md) since all class and animation names are scoped locally by default. This means that it's now possible to use semantic and shorter names for our CSS classes.

So instead of importing a stylesheet in this massive [`_components.scss`](https://github.com/Automattic/wp-calypso/blob/master/assets/stylesheets/_components.scss):

```javascript
@import 'components/vertical-nav/style';
```

And using a class name that must remain unique across Calypso:

```javascript
export default ( { children } ) => {
  return (
    <div className="vertical-nav">
      { children }
    </div>
  );
};
```

We can now just import the stylesheet from the component itself, as if it was a Javascript module:

```javascript
import style from './style';

export default ( { children } ) => {
  return (
    <div className={ style.content }>
      { children }
    </div>
  );
};
```

While still using plain good old CSS, but with shorter class names:

```css
.content {
  margin-top: 15px;
}
```

Webpack takes care of all the magic using three new [loaders](https://webpack.github.io/docs/loaders.html) and preprocesses SCSS files as you import them. To be more specific, the [CSS loader](https://github.com/webpack/css-loader) makes it possible to require CSS files from any component:

```javascript
import style from 'style.css';
```

The [SASS loader](https://github.com/jtangelder/sass-loader) unsurprisingly adds supports for `.scss` files:

```javascript
import style from 'style.scss';
```

Finally, the [Style loader](https://github.com/webpack/style-loader) adds the styles of a component to the `<head>` of the page. A prefix based on the path of the component is automatically injected in front of class names, thus preventing collisions:

```html
<html>
  <head>
    ...
    <style type="text/css">
      .client-components-vertical-nav-content {
        margin-top: 15px;
      }
    </style>
  </head>
  ...
```

#### Other benefits

* We can continue to use `@import` and `url()` in stylesheets since [they will be resolved by Webpack loaders](https://github.com/webpack/css-loader#options) (although all linked assets [must be relative](https://github.com/jtangelder/sass-loader#problems-with-url)).

* We can still target class names that have a global scope with the [`:global()` and `:global {}` constructs](https://github.com/webpack/css-loader#css-scope)

* We can still use composition, either by requiring different stylesheets from a component, or by using the [`composes` keyword](https://github.com/css-modules/css-modules#composition) from the stylesheets themselves (note that when composing multiple classes from different files the order of appliance is undefined).

* We only load CSS the first time a component is mounted.

* We can convert components to use CSS modules gradually.

* We can enable [source maps](https://github.com/jtangelder/sass-loader#source-maps) if we want to.

#### Testing instructions
 
1. Run `git checkout add/css-modules` and start your server, or open a [live branch](https://calypso.live/?branch=add/css-modules)
2. Log into a test account with at least a domain purchased
3. Open the [`Domains` page](http://calypso.localhost:3000/domains/manage)
4. Click the domain name to access the corresponding `Domain Settings` page
5. Check that navigation links are [correctly displayed](https://cloud.githubusercontent.com/assets/594356/21932008/b14d6824-d99e-11e6-864a-003fe1bcea9b.png) and can be clicked

#### Reviews

- [x] Code
- [ ] Product
- [x] Tests

### CODE REVIEW

**Comment by stephanethomas** [comment]
I'm currently encountering the following issue in the [test runner](https://github.com/Automattic/wp-calypso/blob/9865d1bd9a34b960c96be5fa1c0f75b1c9a474ed/test/runner.js#L50):

```
npm run test-client client/my-sites/upgrades/domain-management/edit

> wp-calypso@0.17.0 test-client /var/sources
> NODE_ENV=test NODE_PATH=test:client TEST_ROOT=client test/runner.js "client/my-sites/upgrades/domain-management/edit"

  my-sites
    upgrades
      domain-management
        edit
          mapped-domain
            1) "before all" hook

  0 passing (1s)
  1 failing

  1) my-sites upgrades domain-management edit mapped-domain "before all" hook:
     Error: Cannot find module './style'
      at Function.hookedLoader [as _load] (node_modules/mockery/mockery.js:111:12)
      at require (internal/module.js:20:19)
      at Object.<anonymous> (index.jsx:9:1)
      at loader (node_modules/babel-register/lib/node.js:158:5)
      at Object.require.extensions.(anonymous function) [as .jsx] (node_modules/babel-register/lib/node.js:168:7)
```

**Comment by gziolo** [comment]
No longer an issue with the commit I have just pushed :) However it would be nice to wrap this tweak with some kind of helper to make it easier to use for other developers.

**Comment by blowery** [comment]
> We only load CSS when we need it. Webpack takes care of removing the styles injected in the <head> as soon as a component is unmounted.

How does that work? I'm not seeing any kind of React integration? Or did you mean something else?

**Comment by stephanethomas** [comment]
> How does that work? I'm not seeing any kind of React integration? Or did you mean something else?

You're right @blowery, this is actually a specific feature of the [Isomorphic Style loader](https://github.com/kriasoft/isomorphic-style-loader) that was part of https://github.com/Automattic/wp-calypso/pull/10366. I'll update the pull request description accordingly.

**Comment by blowery** [comment]
@stephanethomas ahhhhh, ok. That makes sense. Sweet! This looks amazing!

File: client/extensions/hello-dolly/index.js [hunk 1]
```diff
@@ -52,7 +52,7 @@ const render = ( context ) => {
 				/>
 			</FAQ>
 		</Main>
-	), document.getElementById( 'primary' ), context.store );
+	), document.getElementById( 'primary' ), context );
```
Comment (gziolo) [inline client/extensions/hello-dolly/index.js]: `document.getElementById( 'primary' )` can be expressed as `primary`.

File: client/extensions/README.md [hunk 2]
```diff
@@ -47,11 +47,11 @@ At the moment we use a simple routing interface with `page.js`. There are a few
 
 ```js
 const renderHelloWorld = ( context ) => {
-	renderWithReduxStore( (
+	renderPage( (
 		<Main>
 			<HelloWorld />
 		</Main>
-	), document.getElementById( 'primary' ), context.store );
+	), document.getElementById( 'primary' ), context );
```
Comment (gziolo) [inline client/extensions/README.md]: document.getElementById( 'primary' ) can be expressed as primary.

File: client/lib/react-helpers/index.js [hunk 3]
```diff
@@ -9,15 +9,25 @@ import { Provider as ReduxProvider } from 'react-redux';
...
 
 export function concatTitle( ...parts ) {
 	return parts.join( ' › ' );
 }
 
-export function renderPage( context, component ) {
-	renderWithReduxStore(
-		component,
-		document.getElementById( 'primary' ),
+export function renderPage( component, domContainer, context ) {
```
Comment (gziolo) [inline client/lib/react-helpers/index.js]: Wouldn't it be easier to add `omContainer` as the last param?

```js
export function renderPage( component, context, domContainer = 'primary' ) {
...
}
```

File: server/pages/index.js [hunk 4]
```diff
@@ -134,7 +134,14 @@ function getDefaultContext( request ) {
 		isFluidWidth: !! config.isEnabled( 'fluid-width' ),
 		abTestHelper: !! config.isEnabled( 'dev/test-helper' ),
 		devDocsURL: '/devdocs',
-		store: createReduxStore()
+		store: createReduxStore(),
+		css: new Set(),
+		insertCss: function() {
```
Comment (gziolo) [inline server/pages/index.js]: Can we use ES6 rest params here?

```js
insertCss( ...styles ) {
 ...
}
```

**Comment by gziolo** [comment]
This is looking really good. I think we should aim to merge it early next week! 💯 

I don't have any comments regarding CSS modules configuration. It's similar to what we have in Automattic/delphin. 

I think we should refactor further `renderPage` method, but it can be done in another PR. Let me explain how I see it.

**Comment by gziolo** [comment]
@lancewillett @alisterscott Is it possible to run e2e tests on this particular branch? :) It's touching lots of page controllers, so it would be nice to double check if it doesn't break anything by accident. To be honest I don't expect any issues, but it wouldn't harm to run full test suite 😃

File: client/lib/react-helpers/index.js (hunk 3 above)
Comment (Tug) [inline client/lib/react-helpers/index.js]: Yes it would!
My goal here was to change the least amount of code possible, doing search and replace only.
Making the optional parameter the last would indeed be a cleaner solution, I'll find the time to update the PR but it's going to be a lot of manual changes :)

Comment (alisterscott) [inline client/lib/react-helpers/index.js]: I might wait til you do this until we run the e2e tests against it. I'm on weekend now (Australia time) but @hoverduck should be able to help out Friday US time

**Comment by alisterscott** [comment]
> @lancewillett @alisterscott Is it possible to run e2e tests on this particular branch? :

I responded above, we can definitely help out, but we might wait until the order of parameters is changed to avoid running them a couple of times.

File: client/lib/react-helpers/index.js (hunk 3 above)
Comment (stephanethomas) [inline client/lib/react-helpers/index.js]: @gziolo, @tug: maybe we should make that improvement in another pull request? This one is already quite long.

Comment (gziolo) [inline client/lib/react-helpers/index.js]: Sure, we can open a follow-up. See also this comment #issuecomment-277193821. I think it can be refactored even further.

File: server/pages/index.js (hunk 4 above)
Comment (stephanethomas) [inline server/pages/index.js]: What about using arrow functions as well? :)

**Comment by scruffian** [comment]
I've tested this and it seems to work perfect 👍

File: npm-shrinkwrap.json [hunk 5]
```diff
@@ -1461,6 +1518,14 @@
     "extglob": {
       "version": "0.3.2"
     },
+    "extract-text-webpack-plugin": {
```
Comment (stephanethomas) [inline npm-shrinkwrap.json]: We should regenerate that file since we no longer use that module.

**Comment by Tug** [comment]
Comments addressed, rebased, I think we're ready for an e2e test run before we merge this @alisterscott :)

**Comment by Tug** [comment]
> I think we should refactor further renderPage method, but it can be done in another PR. Let me explain how I see it. First of all I'm probably guilty of this connection with page in the method name, too :) I don't think it fit well with what we are tying to solve here. My proposal would be to refactor renderPage as follows:
> ...

I would rather leave it as it is.
One of the good refactors we want is to have only one root component and get rid of `primary` and `secondary`. Those functions (`renderPage` and `renderWithReduxStore`) are helpers only and aren't meant to be calypso framework specific functions one must call to render something. They're just here to avoid code duplication.
Adding `renderPrimaryComponent ` and `renderSecondaryComponent` would make `primary` and `secondary` legitimate as part of the application and everyone is going to have to use them and we will just make it harder for everyone to transition into this one root only calypso.

**Comment by hoverduck** [comment]
I ran the e2e tests and had one issue - the secure checkout page never loads.  No errors in the console log, just a blank page.  I can see what's in my shopping cart, but don't get the form to fill out with my payment information.

screen shot 2017-02-03 at 10 30 12 am

**Comment by Tug** [comment]
@hoverduck thanks for running it, seems like it how itself very useful! I can reproduce the bug locally!
Moving this back to `In Progress`

File: client/my-sites/upgrades/controller.jsx [hunk 6]
```diff
@@ -199,18 +195,16 @@ module.exports = {
...
-		renderWithReduxStore(
+		renderPage(
 			(
 				<CartData>
 					<SecondaryCart selectedSite={ sites.getSelectedSite() } />
 				</CartData>
 			),
-			document.getElementById( 'secondary' ),
-			context.store
+			context
```
Comment (Tug) [inline client/my-sites/upgrades/controller.jsx]: Oops seems like I missed this one!

**Comment by Tug** [comment]
Ok fixed it, I did a mistake while refactoring and removed a `'secondary'` param at some point.

File: server/pages/index.js [hunk 7]
```diff
@@ -134,7 +134,13 @@ function getDefaultContext( request ) {
 		isFluidWidth: !! config.isEnabled( 'fluid-width' ),
 		abTestHelper: !! config.isEnabled( 'dev/test-helper' ),
 		devDocsURL: '/devdocs',
-		store: createReduxStore()
+		store: createReduxStore(),
+		css: new Set(),
+		insertCss: ( ...styles ) => {
+			styles.forEach( function( style ) {
+				context.css.add( style._getCss() )
```
Comment (drewblaisdell) [inline server/pages/index.js]: For my own understanding: why do we need to use a `Set` here?

File: server/pages/index.jade [hunk 8]
```diff
@@ -54,6 +54,8 @@ html(lang=lang, dir=isRTL ? 'rtl' : 'ltr', class=isFluidWidth ? 'is-fluid-width'
 				link(rel='stylesheet', href=urls['style-debug.css'])
 			else
 				link(rel='stylesheet', href=urls['style.css'])
+		style(type='text/css' id='critical-path-css')
+			!= Array.from(css).join('')
```
Comment (drewblaisdell) [inline server/pages/index.jade]: Super minor, but there are some spaces missing here.

**Comment by drewblaisdell** [comment]
LGTM and works as expected. I had one question in the notes, and also am curious if there are plans to update the docs here or if we'll do that in a separate PR?

Hopefully @hoverduck can run the e2e tests once more before this is merged. :)

**Comment by alisterscott** [comment]
I've re-run the e2e tests against the latest changes and all passed 👍

**Comment by alisterscott** [comment]
FYI this PR stopped the Jetpack plugin details page from loading in Production due to the incorrect order of the parameters. This was found by the Jetpack e2e tests, I raised this as #11184 and fixed immediately/deployed with the help of @markryall in #11185. Unfortunately I didn't think to run the Jetpack e2e tests against this PR but now we know...

**Comment by Tug** [comment]
Thanks a lot for fixing this @alisterscott !!
I really thought I caught all `renderPage` here, guess I didn't 👍

**Comment by ehg** [comment]
Can we revert until we've had a framework discussion surrounding this? I'm worried that we may adversely affect caching here. I'd prefer if we could investigate what impact this would have Calypso-wide, before others follow this pattern… /cc @mtias

**Comment by kwight** [comment]
FWIW, after a rebase this morning, I've started getting these errors, which sound like they might be related:

screen shot 2017-02-06 at 1 07 52 pm

This is for #10542 , which uses `renderPage()` [here](https://github.com/Automattic/wp-calypso/pull/10542/files#diff-5ce7e1b12afe6d1a309f9f291f36e997R191).

**Comment by mtias** [comment]
Yes, please, let's revert and discuss this a bit more. It's a big change: there are caching implications, class names losing their readability, interactions with tools like RTL formatting, impact on server / build time, composition uncertainties (`.site .site-icon` example), that we need to go over before committing to it.

**Comment by ockham** [comment]
I'll add to what @mtias said that if we're redoing routing on that large a scale, I'd love if we did in a more future-proof/isomorphic way. Another thing to discuss 😄

File: server/pages/index.js (hunk 7 above)
Comment (Tug) [inline server/pages/index.js]: I guess because the same component can be mounted several times and we don't want to have too many duplications of our css styles.

File: client/lib/react-helpers/stylizer.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import { Children, Component, PropTypes } from 'react';
```
Comment (scruffian) [inline client/lib/react-helpers/stylizer.js]: Trivial, but I guess we should have one of these here:
```/**
 * External dependencies
 */```

Comment (scruffian) [inline client/lib/react-helpers/stylizer.js]: Whoops commented on the wrong PR
